libcontainer/state_linux.go

```diff
@@ -0,0 +1,218 @@
+// +build linux
+
+package libcontainer
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/opencontainers/runc/libcontainer/configs"
+)
+
+func newStateError(from, to containerState) error {
+	return &stateError{
+		From: from.status().String(),
+		To:   to.status().String(),
+	}
+}
+
+// stateError is returned when an invalid state transition happens from one
+// state to another.
+type stateError struct {
+	From string
+	To   string
+}
+
+func (s *stateError) Error() string {
+	return fmt.Sprintf("invalid state transition from %s to %s", s.From, s.To)
+}
+
+type containerState interface {
+	transition(containerState) error
+	destroy() error
+	status() Status
+}
+
+func destroy(c *linuxContainer) error {
+	if !c.config.Namespaces.Contains(configs.NEWPID) {
+		if err := killCgroupProcesses(c.cgroupManager); err != nil {
+			logrus.Warn(err)
+		}
+	}
+	err := c.cgroupManager.Destroy()
+	if rerr := os.RemoveAll(c.root); err == nil {
+		err = rerr
+	}
+	c.initProcess = nil
+	if herr := runHooks(c); err == nil {
+		err = herr
+	}
+	return err
+}
+
+func runHooks(c *linuxContainer) error {
```

Comment:
this func has pretty confusing name, considering that it's located in libcontainer package.